OpenRA.Mods.Common/DiscordService.cs

```diff
@@ -0,0 +1,194 @@
+﻿#region Copyright & License Information
+/*
+ * Copyright 2007-2020 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System;
+using DiscordRPC;
+using DiscordRPC.Message;
+using OpenRA.Network;
+
+namespace OpenRA.Mods.Common
+{
+	public enum DiscordState
+	{
+		Unknown,
+		InMenu,
+		InMapEditor,
+		InSinglePlayer,
+		InMultiplayerLobby,
+		PlayingMultiplayer,
+		WatchingReplay,
+		PlayingCampaign
+	}
+
+	public class DiscordService : IGlobalModData, IDisposable
+	{
+		public readonly string ApplicationId = null;
+		readonly DiscordRpcClient client;
+		DiscordState currentState;
+
+		static readonly Lazy<DiscordService> Service = Exts.Lazy(() =>
+		{
+			if (!Game.ModData.Manifest.Contains<DiscordService>())
+				return null;
+
+			return Game.ModData.Manifest.Get<DiscordService>();
+		});
+
+		public DiscordService(MiniYaml yaml)
+		{
+			FieldLoader.Load(this, yaml);
+
+			client = new DiscordRpcClient(ApplicationId, autoEvents: false)
+			{
+				SkipIdenticalPresence = false
+			};
+
+			client.OnJoin += OnJoin;
+			client.OnJoinRequested += OnJoinRequested;
+
+			// Hack: We need to bypass the check that checks that HasRegisteredUriScheme is set
```

Comment:
style nit: `// HACK:`. Would be good if we could expand the detail here a bit to explain that the library expects to register the uri handler itself, but that we are doing that ourselves in our installer/launcher.